File: app/lib/data/repositories/day_task_repository.dart
import 'dart:convert';
import 'package:drift/drift.dart';
import '../db/database.dart';
import '../../domain/models/day_task.dart';
import '../../domain/models/exercise.dart';
import '../../domain/models/meal.dart';

class DayTaskRepository {
  final AppDatabase db;
  DayTaskRepository(this.db);

  Future<DayTask?> getByDate(int planId, DateTime date) async {
    final row = await (db.select(db.dayTasks)
          ..where((t) => t.planId.equals(planId) & t.date.equals(date))
          ..limit(1))
        .getSingleOrNull();
    if (row == null) return null;
    return _toModel(row);
  }

  Future<List<DayTask>> getAllForPlan(int planId) async {
    final rows = await (db.select(db.dayTasks)
          ..where((t) => t.planId.equals(planId))
          ..orderBy([(t) => OrderingTerm.asc(t.dayIndex)]))
        .get();
    return rows.map(_toModel).toList();
  }

  Future<void> markWorkoutDone(int id) async {
    await (db.update(db.dayTasks)..where((t) => t.id.equals(id))).write(
      DayTasksCompanion(
        completedWorkout: const Value(true),
        completedAt: Value(DateTime.now()),
      ),
    );
  }

  Future<void> markMealDone(int id, String slot, bool done) async {
    final row = await (db.select(db.dayTasks)..where((t) => t.id.equals(id))).getSingle();
    final map = (jsonDecode(row.completedMeals) as Map).cast<String, bool>();
    map[slot] = done;
    await (db.update(db.dayTasks)..where((t) => t.id.equals(id))).write(
      DayTasksCompanion(completedMeals: Value(jsonEncode(map))),
    );
  }

  Future<void> awardXp(int id, int additionalXp) async {
    final row = await (db.select(db.dayTasks)..where((t) => t.id.equals(id))).getSingle();
    await (db.update(db.dayTasks)..where((t) => t.id.equals(id))).write(
      DayTasksCompanion(xpAwarded: Value(row.xpAwarded + additionalXp)),
    );
  }

  DayTask _toModel(DayTaskRow row) {
    final workoutJson = row.workoutJson.isEmpty ? null : jsonDecode(row.workoutJson);
    final mealsJson = (jsonDecode(row.mealsJson) as List).cast<Map<String, dynamic>>();
    return DayTask(
      dbId: row.id,
      planId: row.planId,
      dayIndex: row.dayIndex,
      date: row.date,
      workout: workoutJson == null
          ? null
          : Workout(
              title: workoutJson['title'] as String,
              estimatedMinutes: workoutJson['estimated_minutes'] as int,
              exercises: (workoutJson['exercises'] as List)
                  .cast<Map<String, dynamic>>()
                  .map(Exercise.fromJson)
                  .toList(),
            ),
      meals: mealsJson.map(Meal.fromJson).toList(),
      completedWorkout: row.completedWorkout,
      completedMeals:
          (jsonDecode(row.completedMeals) as Map).cast<String, bool>(),
      xpAwarded: row.xpAwarded,
      completedAt: row.completedAt,
    );
  }
}

File: app/android/app/src/main/java/com/gymcoach/gym_coach/data/repository/DayTaskRepository.kt
package com.gymcoach.gym_coach.data.repository

import com.gymcoach.gym_coach.data.db.dao.DayTaskDao
import com.gymcoach.gym_coach.data.db.entity.DayTaskEntity
import com.gymcoach.gym_coach.domain.model.DayTask
import com.gymcoach.gym_coach.domain.model.Meal
import com.gymcoach.gym_coach.domain.model.Workout
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import java.time.LocalDate
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class DayTaskRepository @Inject constructor(
    private val dao: DayTaskDao
) {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun getByDate(planId: Int, date: LocalDate): DayTask? =
        dao.getByDate(planId.toLong(), date.toEpochDay())?.let(::toModel)

    suspend fun getAllForPlan(planId: Int): List<DayTask> =
        dao.getAllForPlan(planId.toLong()).map(::toModel)

    suspend fun markWorkoutDone(id: Int) =
        dao.markWorkoutDone(id.toLong(), System.currentTimeMillis())

    suspend fun markMealDone(id: Int, slot: String, done: Boolean) =
        dao.markMealDoneAtomic(id.toLong(), slot, done)

    suspend fun awardXp(id: Int, additionalXp: Int) {
        if (additionalXp <= 0) return
        dao.addXp(id.toLong(), additionalXp)
    }

    suspend fun totalXp(): Int = dao.totalXp()
    suspend fun totalWorkoutsCompleted(): Int = dao.totalWorkoutsCompleted()

    /**
     * Mirrors `lib/data/repositories/gamification_repository.dart`:
     * count rows where completed_workout = true AND completed_meals JSON
     * object has every value == true AND the map is non-empty.
     */
    suspend fun totalDaysAllMealsCompleted(): Int {
        // The Drift code filtered in-memory; we do the same by fetching the
        // completed-meals JSON for completed-workout rows and parsing in
        // Kotlin.
        return dao.completedWorkoutRows().count { row ->
            val map = parseCompletedMeals(row.completedMeals)
            map.isNotEmpty() && map.values.all { it }
        }
    }

    private fun toModel(row: DayTaskEntity): DayTask {
        val workout = if (row.workoutJson.isEmpty()) null
        else json.decodeFromString(Workout.serializer(), row.workoutJson)
        val meals = if (row.mealsJson.isEmpty()) emptyList()
        else json.decodeFromString(ListSerializer(Meal.serializer()), row.mealsJson)
        val completed = parseCompletedMeals(row.completedMeals)
        return DayTask(
            dbId = row.id.toInt(),
            planId = row.planId.toInt(),
            dayIndex = row.dayIndex,
            date = LocalDate.ofEpochDay(row.date),
            workout = workout,
            meals = meals,
            completedWorkout = row.completedWorkout,
            completedMeals = completed,
            xpAwarded = row.xpAwarded,
            completedAt = row.completedAt
        )
    }

    private fun parseCompletedMeals(raw: String): Map<String, Boolean> {
        if (raw.isEmpty() || raw == "{}") return emptyMap()
        return runCatching {
            json.decodeFromString(MapSerializer(String.serializer(), Boolean.serializer()), raw)
        }.getOrElse { emptyMap() }
    }
}
